node/src/components/consensus/highway_core/state.rs

```diff
@@ -387,11 +389,19 @@ impl<C: Context> State<C> {
                         return Err(VoteError::RoundLength);
                     }
                 }
+                let r_id = round_id(wvote.timestamp, wvote.round_exp);
                 if let Some(prev2_vote) = prev_vote.previous().map(|h2| self.vote(h2)) {
-                    if prev2_vote.round_id() == round_id(wvote.timestamp, wvote.round_exp) {
+                    if prev2_vote.round_id() == r_id {
                         return Err(VoteError::ThreeVotesInRound);
                     }
                 }
+                if wvote.value.is_some()
+                    && (prev_vote.round_id() == r_id
+                        || wvote.timestamp != r_id
+                        || self.leader(r_id) != creator)
+                {
+                    return Err(VoteError::NonLeaderBlock);
```

Comment:
I'm surprised we have not been checking this earlier (I would have sworn we have).